lib/src/fnc/crypto.rs

```diff
@@ -39,23 +39,72 @@ pub fn sha512(_: &Context, mut args: Vec<Value>) -> Result<Value, Error> {
 	Ok(val.into())
 }
 
+/// Allowed to cost this much more than default setting for each hash function.
+const COST_ALLOWANCE: u32 = 4;
+
+/// Like verify_password, but takes a closure to determine whether the cost of performing the
+/// operation is not too high.
+macro_rules! bounded_verify_password {
+	($algo: ident, $instance: expr, $password: expr, $hash: expr, $bound: expr) => {
+		if let (Some(salt), Some(expected_output)) = (&$hash.salt, &$hash.hash) {
+			if let Some(params) =
+				<$algo as PasswordHasher>::Params::try_from($hash).ok().filter($bound)
+			{
+				if let Ok(computed_hash) = $instance.hash_password_customized(
+					$password.as_ref(),
+					Some($hash.algorithm),
+					$hash.version,
+					params,
+					*salt,
+				) {
+					if let Some(computed_output) = &computed_hash.hash {
+						expected_output == computed_output
+					} else {
+						false
+					}
+				} else {
+					false
+				}
+			} else {
+				false
+			}
+		} else {
+			false
+		}
+	};
+
+	($algo: ident, $password: expr, $hash: expr, $bound: expr) => {
+		bounded_verify_password!($algo, $algo::default(), $password, $hash, $bound)
+	};
+}
+
```

Comment:
Need macro since each hash function has a separate import path for the `password_hash` crate, so using a trait wouldn't work.